error.go

```diff
@@ -292,6 +293,12 @@ func wrapIfContextError(err error) error {
 	if errors.Is(err, context.DeadlineExceeded) {
 		return NewError(CodeDeadlineExceeded, err)
 	}
+	// Ick, some dial errors can be returned as os.ErrDeadlineExceeded
+	// instead of context.DeadlineExceeded :(
+	// https://github.com/golang/go/issues/64449
+	if errors.Is(err, os.ErrDeadlineExceeded) {
+		return NewError(CodeDeadlineExceeded, err)
+	}
```

Comment:
Confirming: whenever we hit this case, the actual context has been cancelled - correct?

Reply:
Okay, I think I've now squashed _all_ of the bugs.

The upside is that the test seemed to fail pretty consistently without all of these fixes, so I think it will be a reasonably good way to identify issues in CI. But there are a couple of downsides:
1. It never repro'ed the issues with the race detector enabled. So I have the test currently running only when the race detector is disabled. And so there is now an extra command run for `make test` that just runs this one test, without `-race`.
2. When there are no bugs, it slows down CI by a full 20 seconds, hoping to find a non-existent bug.

See the last three commits for details:
1.  d5965e8f21e308b2ce9b9fefc868a763f63634cd adds the test
2. 39a162c3ded6549ab0379ab2f40e371b8883b86d updates the test so the output is more useful in tracking down the source of the error; this also makes sure to use the gRPC protocol, too, in case there protocol-specific places that fail to classify the error correctly. (It just uses gRPC-Web right now, since that is the same protocol handler implementation and basically a superset of activity, since it includes add'l body writes for the trailers.)
3. d42edba030949fe950ffc486fb7eb5570f7d49e3 fixes the three add'l places found by the test that were incorrectly classifying errors. One of them was wrapping a `*connect.Error` with another `*connect.Error` and changing the code to "unknown". Oops!

Reply:
> One of them was wrapping a *connect.Error with another *connect.Error and changing the code to "unknown". Oops!

😬 Yikes!